### PR TITLE
API: add support for sklearn clustering algorithms

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @j-ittner @joerg-schneider!

Thank you for your guidance!

Here's a work-in-progress PR with clusterers support, for which I followed existing implementation. Please review if that's the right direction to go.

Here all clusterers are supported that implement `ClusterMixin`. For tomorrow I plan to add separate internal `clustering` wrappers for two `KMeans` methods to also wrap `cluster_centers_` in a dataframe.

Also I've got a few questions about this design. One of them is that extending `LearnerDF` by `ClustererDF` imposes presence of `score` method which is available in e.g. `KMeans` but not in `AgglomerativeClustering`. Other approach would be to remove `score` from `LearnerDF` and introduce `LearnerWithScoringDF` in the inheritance hierarchy but for the first POC I wanted to modify as little existing codebase as possible and wait for your review.

Here's short interaction with this implementation: `labels_` preserves indexes and columns of DF used for fitting, also `predict` does:

```python3
>>> import sklearn
>>> from sklearn.cluster import KMeans
>>> from sklearndf.clustering import KMeansDF
>>> import numpy as np
>>> import pandas as pd
>>>
>>> data = np.array([[1, 1], [1, 4], [1, 0], [10, 3], [10, 4], [10, 0]])
>>> columns = ['bmi', 'age']
>>> index = ['a', 'b', 'c', 'd', 'e', 'f']
>>> X = pd.DataFrame(data=data, columns=columns, index=index)
>>> X_test = pd.DataFrame(data=[[0, 0], [5, 5]], columns=columns, index=['t', 's'])
>>>
>>> kmeans = KMeans(n_clusters=2, random_state=42).fit(X)
>>> kmeans.labels_
array([1, 1, 1, 0, 0, 0], dtype=int32)
>>> kmeans.predict(X_test)
array([1, 1], dtype=int32)
>>>
>>> kmeansdf = KMeansDF(n_clusters=2, random_state=42).fit(X)
>>> kmeansdf.labels_
a    1
b    1
c    1
d    0
e    0
f    0
Name: labels, dtype: int32
>>> kmeansdf.predict(X_test)
t    1
s    1
Name: prediction, dtype: int32
```

### CODE REVIEW

File: src/sklearndf/clustering/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+"""
+Extended versions of Scikit-Learn clusterers with enhanced E2E support for data
+frames.
+"""
+from .. import __sklearn_0_22__, __sklearn_0_23__, __sklearn_version__
```
Comment (j-ittner) [inline src/sklearndf/clustering/__init__.py]: We don't need (and should not) import these - this is done elsewhere (and the _x_xx__ packages are conditional imports to ensure backwards compatibility to earlier versions of scikit-learn)

**Review by j-ittner — COMMENTED** [review]
Fantastic start @mtsokol!

On your question: I would base `ClustererDF` on `EstimatorDF` so we get the following top-level class hierarchy:
```
EstimatorDF
    TransformerDF
    LearnerDF
        RegressorDF
        ClassifierDF
    ClustererDF
```

File: src/sklearndf/_sklearndf.py [hunk 2]
```diff
@@ -416,4 +424,21 @@ def decision_function(
         """
 
 
+class ClustererDF(LearnerDF, ClusterMixin, metaclass=ABCMeta):
```
Comment (j-ittner) [inline src/sklearndf/_sklearndf.py:498]: Base on `EstimatorDF` not `LearnerDF`

File: src/sklearndf/_sklearndf.py [hunk 3]
```diff
@@ -416,4 +424,21 @@ def decision_function(
...
+    Provides enhanced support for data frames.
+    """
+
+    @property
+    @abstractmethod
+    def labels_(self) -> pd.Series:
+        """
+        A pandas series, mapping input dataframe that was used for fitting to the
+        specific cluster's label (decided by a clustering algorithm).
+        """
```
Comment (j-ittner) [inline src/sklearndf/_sklearndf.py]: Not sure this is entirely clear: is this really a mapping? How are these values determined by the algorithm?

We may not need this property at all, not all clusterers implement this and not sure we need to provide a conversion to a pandas series (we are not mapping feature names to anything here)

File: src/sklearndf/_sklearndf.py (hunk 2 above)
Comment (mtsokol) [reply src/sklearndf/_sklearndf.py:498]: [EDIT] Moving discussion to this comment

Sure, sounds great! I've got one follow up question on this design. 
`LearnerDF` also provides `predict` and `fit_predict` methods for wrapping that are used by all clusterers. 

I can either:
1. Make `ClustererDF` inherit from `EstimatorDF` but then I need to repeat both `predict` and `fit_predict` methods implementations there.
2. Add separate `ScorerDF` with only `score` method that will be inherited by Regressor and Classifier (in the second place considering multiple inheritance here: `RegressorDF(LearnerDF, ScorerDF)`) and remove `score` from `LearnerDF`.
~Would it be OK to go with the second one? Or stick to the first one?~ [OUTDATED]

Comment (j-ittner) [reply src/sklearndf/_sklearndf.py:498]: I would not base clusterers on `LearnerDF`.
1. semantically, they are not supervised learners in the scikit-learn terminology
2. not all clusterers implement `predict`, e.g. `DBSCAN`
3. `ScorerDF` makes the class hierarchy unnecessarily complex
4. you need to add abstract method `fit_predict` to `ClustererDF` (but with a different signature than in `LearnerDF`, to reflect that arg `y` is optional.

File: src/sklearndf/clustering/__init__.py (hunk 1 above)
Comment (mtsokol) [reply src/sklearndf/clustering/__init__.py]: Sure, removed!

File: src/sklearndf/_sklearndf.py (hunk 2 above)
Comment (mtsokol) [reply src/sklearndf/_sklearndf.py:498]: Hi @j-ittner, closing note for today: So based on our today's discussion - an updated inheritance hierarchy would look like this (I've got it partially completed) which I can file as a separate refactor PR to `2.0.x` branch, which after merging will make incorporating Clusterers possible in an easy way. Is this OK?

```
                 ┌───────────┐
            ┌───►│EstimatorDF│  # introduces `fit`
            │    └───────────┘
            │          ▲
            │          │
 ┌──────────┴──┐  ┌────┴────┐
 │TransformerDF│  │LearnerDF│  # introduces `predict` and `fit_predict`
 └─────────────┘  └─────────┘
                   ▲      ▲
         ┌─────────┘      │
         │                │
┌────────┴──┐   ┌─────────┴─────────┐
│ClustererDF│   │SupervisedLearnerDF│ # introduces `score`
└───────────┘   └───────────────────┘
                 ▲               ▲
                 │               │
                 │               │
      ┌──────────┴─┐    ┌────────┴──┐
      │ClassifierDF│    │RegressorDF│
      └────────────┘    └───────────┘
```

Comment (mtsokol) [reply src/sklearndf/_sklearndf.py:498]: After introducing #194 now `ClustererDF` can inherit from `LearnerDF`.

**Comment by mtsokol** [comment]
Hi @j-ittner, I rebased it to the latest `2.0.x` and now inheritance is fixed. 
PR is ready for another round of review!

File: src/sklearndf/_sklearndf.py (hunk 3 above)
Comment (j-ittner) [reply src/sklearndf/_sklearndf.py]: Ah got it - this is not the list of unique labels, but we are mapping the row index of the ingoing data frame to the label for each observation. Suggest you clarify the docstring a bit.

File: src/sklearndf/_sklearndf.py (hunk 2 above)
Comment (j-ittner) [reply src/sklearndf/_sklearndf.py:498]: This is a great solution. We need one more change in class `LearnerDF`: arg `y` of method `LearnerDF.fit_predict()` should have signature `Optional[Union[pd.Series, pd.DataFrame]]`, in analogy with the signature of method `EstimatorDF.fit()`

File: src/sklearndf/wrapper/_wrapper.py [hunk 4]
```diff
@@ -824,6 +829,60 @@ def _prediction_with_class_labels(
...
+        raw_labels = self._native_estimator.labels_
+
+        return pd.Series(data=raw_labels, name=self.COL_LABELS, index=self.X_index)
+
+    def fit(
+        self: T_Self,
+        X: pd.DataFrame,
+        y: Optional[Union[pd.Series, pd.DataFrame]] = None,
+        **fit_params: Any,
+    ) -> T_Self:
```
Comment (j-ittner) [inline src/sklearndf/wrapper/_wrapper.py]: Instead of overriding method `fit`, I suggest you override these `EstimatorDF` methods:

- `_check_parameter_types` to ensure that `y is None`
- `_post_fit` to preserve row labels (please use a protected field and lower case, e.g., `_x_index`)
- `_reset_fit` to reset `_x_index` to `None`

File: test/test/sklearndf/test_sklearn_coverage.py [hunk 5]
```diff
@@ -57,11 +59,20 @@
     *GENERAL_COVERAGE_EXCLUSIONS,
     # class "Imputer" was deprecated in 0.20 and removed in 0.22
     "Imputer",
+    "AgglomerationTransform",  # TODO should be?
```
Comment (j-ittner) [inline test/test/sklearndf/test_sklearn_coverage.py]: I agree this should not be wrapped, given it is a mix-in class. Could you please add a comment in the line above with a short explanation for the exclusion? Similar to how we explain why we exclude `Imputer`.

File: test/test/sklearndf/test_sklearn_coverage.py [hunk 6]
```diff
@@ -57,11 +59,20 @@
...
 }
 
 
 PIPELINE_COVERAGE_EXCLUSIONS = GENERAL_COVERAGE_EXCLUSIONS
 
+
+CLUSTERER_COVERAGE_EXCLUSIONS = {
+    *GENERAL_COVERAGE_EXCLUSIONS,
+    # class "FeatureAgglomeration" is already included via Transformers
+    "FeatureAgglomeration",
```
Comment (j-ittner) [inline test/test/sklearndf/test_sklearn_coverage.py:74]: Shouldn't we also test `FeatureAgglomeration` as part of the clustering tests? We want to confirm that FeatureAgglomeration also implements `ClustererDF`.

File: test/test/sklearndf/test_sklearn_coverage.py [hunk 7]
```diff
@@ -196,3 +216,16 @@ def test_pipeline_coverage(sklearn_pipeline_cls: Type) -> None:
 
     if sklearn_pipeline_cls not in sklearn_classes:
         _check_unexpected_sklearn_class(sklearn_pipeline_cls)
+
+
+@pytest.mark.parametrize(
+    argnames="sklearn_clusterer_cls", argvalues=sklearn_clusterer_classes()
+)
+def test_clusterer_coverage(sklearn_clusterer_cls: Type[ClusterMixin]) -> None:
```
Comment (j-ittner) [inline test/test/sklearndf/test_sklearn_coverage.py:226]: It's great that we have the coverage test also for clusterers. Could you please also add unit tests that call `fit` and `fit_predict` on some or all of the clusterers, so we have a test for the wrapper class?

**Review by j-ittner — CHANGES_REQUESTED** [review]
Great progress!

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 6 above)
Comment (mtsokol) [reply test/test/sklearndf/test_sklearn_coverage.py:74]: To make tests for clusterers pass then I need to also wrap `FeatureAgglomeration` in clusterer module. But then we will have redundancy as `FeatureAgglomeration` will be present in clusterers and transformers:
```
from sklearndf.transformation import FeatureAgglomerationDF
from sklearndf.clustering import FeatureAgglomerationDF
```

I think we can either leave `FeatureAgglomeration` in transformers as is and keep that exclusion or move `FeatureAgglomeration` to clusterers (but then I need to get `ColumnPreservingTransformerWrapperDF` in clusterers package which is implemented in transformers package that this algorithm uses).

Should I leave it as is or move `FeatureAgglomeration` to clusterers?

File: src/sklearndf/wrapper/_wrapper.py (hunk 4 above)
Comment (mtsokol) [reply src/sklearndf/wrapper/_wrapper.py]: Sure! I did it by implementing `_post_fit` and `_reset_fit`.
For `_check_parameter_types` I don't think that it's necessary to assert on `y is None` as `sklearn` allows users to pass `y` always to clusterers and just ignores that argument (if we want to stick more to sklearn compatibility).
Can I omit assertion on `y is None` for input?

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 7 above)
Comment (mtsokol) [reply test/test/sklearndf/test_sklearn_coverage.py:226]: Sure, I added tests for `fit_predict` and `fit` for clusterers.

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 5 above)
Comment (mtsokol) [reply test/test/sklearndf/test_sklearn_coverage.py]: Sure! I added an explanation for that.

File: src/sklearndf/_sklearndf.py (hunk 3 above)
Comment (mtsokol) [reply src/sklearndf/_sklearndf.py]: Sure! Done

File: src/sklearndf/_sklearndf.py (hunk 2 above)
Comment (mtsokol) [reply src/sklearndf/_sklearndf.py:498]: As discussed `fit_predict` was moved to `ClustererDF`. Also changed signature.

File: src/sklearndf/clustering/wrapper/_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+):
+    """
+    DF wrapper for KMeans-like algorithms, e.g., :class:`sklearn.cluster.KMeans`.
+    """
+
+    @property
+    def cluster_centers_(self) -> pd.DataFrame:
+        """
+        The data frame containing cluster centers designated by a clustering algorithm.
+        """
```
Comment (j-ittner) [inline src/sklearndf/clustering/wrapper/_wrapper.py:57]: Let's state what the rows and columns are in the resulting data frame.

File: src/sklearndf/pipeline/_learner_pipeline.py [hunk 9]
```diff
@@ -218,15 +218,6 @@ def predict(
...
-        self, X: pd.DataFrame, y: pd.Series, **fit_params: Any
-    ) -> Union[pd.Series, pd.DataFrame]:
-        """[see superclass]"""
-        return self.final_estimator.fit_predict(
-            self._pre_fit_transform(X, y, **fit_params), y, **fit_params
-        )
-
 
 @inheritdoc(match="[see superclass]")
 class SupervisedLearnerPipelineDF(
```
Comment (j-ittner) [inline src/sklearndf/pipeline/_learner_pipeline.py:225]: Do you agree we should also define a `ClustererPipelineDF`?

File: src/sklearndf/wrapper/_wrapper.py [hunk 10]
```diff
@@ -824,6 +798,83 @@ def _prediction_with_class_labels(
...
+        **fit_params,
+    ) -> None:
+        super()._post_fit(X, y, **fit_params)
+        self._x_index = X.index
+
+    def _reset_fit(self) -> None:
+        super()._reset_fit()
+        self._x_index = None
+
+    def fit_predict(
```
Comment (j-ittner) [inline src/sklearndf/wrapper/_wrapper.py:838]: Just for cosmetics, I would define public methods first and move the protected ones further down.

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 6 above)
Comment (j-ittner) [reply test/test/sklearndf/test_sklearn_coverage.py:74]: Agree if this is just about testing coverage and not testing functionality such as `fit_predict`.
`FeatureAgglomeration` lives in `sklearn.cluster` so perhaps move it to clusters also in our test code?

**Review by j-ittner — CHANGES_REQUESTED** [review]

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 6 above)
Comment (mtsokol) [reply test/test/sklearndf/test_sklearn_coverage.py:74]: I took another look at it and I think this one is a little bit more complex. So in `sklearn` `FeatureAgglomeration` that is placed in `clustering` implements both `ClusterMixin` and `TransformerMixin` and it's both clusterer and transformer. If I move it to `clustering` package and wrap it in `ClustererDF` it will lost wrapping for `transform` and `inverse_transform` methods. 

My idea is to leave it where it is for this PR and prepare another one where:
1.  `FeatureAgglomeration` will be moved to `clustering` package
2. It will be wrapped in such a way that it will be both `TransformerDF` and `ClustererDF` (I need to see if that's going to be even feasible without too much hacking)

What do you think?

File: src/sklearndf/pipeline/_learner_pipeline.py (hunk 9 above)
Comment (mtsokol) [reply src/sklearndf/pipeline/_learner_pipeline.py:225]: Yes, I added `ClustererPipelineDF` with tests.

File: src/sklearndf/wrapper/_wrapper.py (hunk 10 above)
Comment (mtsokol) [reply src/sklearndf/wrapper/_wrapper.py:838]: Sure, reordered.

File: src/sklearndf/clustering/wrapper/_wrapper.py (hunk 8 above)
Comment (mtsokol) [reply src/sklearndf/clustering/wrapper/_wrapper.py:57]: Sure, I added information about it.

**Comment by mtsokol** [comment]
Hi @j-ittner, I introduced all of the changes from your last review - it's ready for another one.

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 6 above)
Comment (j-ittner) [reply test/test/sklearndf/test_sklearn_coverage.py:74]: Agree, let's handle this in a separate PR. Implementing both `TransformerDF` and `ClustererDF` should be doable; see class `RegressorTransformerWrapperDF` for inspiration (implementing both the `TransformerDF` and `RegressorDF` interface)

**Review by j-ittner — APPROVED** [review]
Looks great, ready to merge!

File: src/sklearndf/wrapper/_wrapper.py (hunk 4 above)
Comment (j-ittner) [reply src/sklearndf/wrapper/_wrapper.py]: If scikit-learn ignores y, it's ok if we do the same. I would not use assertions to check user input, they should only be used to catch unexpected conditions caused by bugs

File: test/test/sklearndf/test_sklearn_coverage.py (hunk 6 above)
Comment (mtsokol) [reply test/test/sklearndf/test_sklearn_coverage.py:74]: @j-ittner Sure! I've created a separate PR for that: #201
